Python transaction handling

kacperpONS opened the pull request:
# Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

This branch adds an implementation of Python transaction handling covered by this [Spike](https://github.com/ONSdigital/ssdc-rm-documentation/pull/234/files#diff-7863fd55e307cb594b4b0daff444928ccba79d08671a07524d3a88bbfd29d2c7)

# What has changed
<!--- What code changes has been made -->
<!--- Has there been any refactoring -->
<!--- What tests have been written -->

Added the transaction handling decorators with appropriate logging

# How to test?
<!--- Describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->
<!--- Are there any automated tests that mean changes don't need to be manually changed -->

Nothing to test yet, still needs to be implemented 
The mechanism has been tested on this docker-dev [branch](https://github.com/ONSdigital/ssdc-rm-docker-dev/tree/1346-spike-transaction-handling-in-python)

# Links
<!--- Add any links to issues (trello, github issues) -->
<!--- Links to any documentation -->
<!--- Links to any related PRs -->
https://jira.ons.gov.uk/secure/RapidBoard.jspa?rapidView=2913&view=detail&selectedIssue=SDCSRM-260

# Screenshots (if appropriate):

AdamHawtin commented on python_spike/caseprocessor/decorators/transaction_decorators.py:
```diff
@@ -0,0 +1,51 @@
+        except Exception as e:
+            logging.error(f"PubSub decorator caugth error: {e}")
+            pubsub_message.nack()
+            logging.info(f"Message nacked: {pubsub_message.data}\n")
```
I think this needs to re-raise the exception here?

kacperpONS replied:
If we raise the exception here, there's currently no way for us to handle it, which would cause a crash. Instead we want to let PubSub nack the message, hoping that next time it arrives we'll be able to process it.

Are you thinking about the case when there's a prolonged problem and we actually want it to crash?

AdamHawtin replied:
Ah, I see I misunderstood.

With the way this is set up, it should probably be combined into one decorator to utilise it, since it would break if you used them in the wrong order or tried to use one on its own? Perhaps that can be a future enhancement instead.

dbanks91 commented on python_spike/caseprocessor/decorators/transaction_decorators.py:
```diff
@@ -0,0 +1,51 @@
+        except Exception as e:
+            Session.remove()
+            logging.error(f"DB transaction decorator caught exception: {e}\n")
+            raise Exception("SQL transaction error")
```
Might be better to pass the error through here too so it's not a generic error message?

kacperpONS replied:
That's a good point, however on second thought the error is already logged once here and passing it like that would lead to duplicating it when caught by @pubsub_transaction.

Seems like another issue that combing the two decorators into one would solve :p

AdamHawtin replied:
I think future versions of these should probably factor out the error handling/logging into its own layer including integration with our exception manager, so I wouldn't worry about perfecting it here and now.

kacperpONS commented:
@AdamHawtin @dbanks91 
I've added an attempt of combining the two decorators into one. I've also updated the code on the docker-dev branch for you to see that it works.
Let me know what you think and I'll remove the two separate decorators.

AdamHawtin commented on python_spike/caseprocessor/decorators/transaction_decorators.py:
```diff
@@ -0,0 +1,84 @@
+    @wraps(callback)
+    def with_transaction_handling(*args, **kwargs):
+        local_session = Session()
+        pubsub_message = args[0]
```
Where it has to interact with the wrapped functions arguments like this, it's probably better to stick to kwargs rather than positional args

kacperpONS replied:
That's what I was thinking, but I wasn't able to find what the keyword argument would be! 

If you look at the [docs](https://cloud.google.com/python/docs/reference/pubsub/latest/google.cloud.pubsub_v1.subscriber.client.Client#google_cloud_pubsub_v1_subscriber_client_Client_subscribe) for the subscribe function and scroll down to the parameters list, you'll see that the callback is passed as an argument to the subscribe function and that the callback function "...receives the message as its only argument...", but there's no more details and the argument name isn't stated.
This suggests the PubSub message is somehow passed under the hood to the callback when it arrives and in my understanding will always be the first argument.

I think that the risk of confusion here is relatively low because we'll only pass the callback to the subscribe function, without explicitly passing any arguments to it.
Also my bad, should've probably left a word on that. Let me know what you think.

AdamHawtin replied:
Makes sense, I forgot our design here is also constrained by what the Pubsub library requires for a callback.

AdamHawtin commented on python_spike/caseprocessor/decorators/transaction_decorators.py:
```diff
@@ -0,0 +1,84 @@
+        local_session = Session()
+        pubsub_message = args[0]
+
+        try:
+            callback(*args, **kwargs, session=local_session)
```
The call to the wrapped function should happen in its own try block, so the error handling can distinguish between exceptions raised in the wrapped function and errors that might happen when the transaction wrapper tries to commit the session or ack the message.

kacperpONS replied:
Good spot, but wouldn't it be clearer if this error handling was happening in the callback function itself?

AdamHawtin replied:
The callback function will need to do any of its own specific error handling, but ultimately for this pattern to work it's going to have to just raise any errors up, so the transaction wrapper can know to respond.

AdamHawtin reviewed (approved):
Looks like an excellent starting point, it would be good to next follow this up by updating the current prototype to use this pattern.
